src/tigerbeetle/scrub.zig

```diff
@@ -0,0 +1,510 @@
+
+    scrub.buffer_prepare = try gpa.alignedAlloc(
+        u8,
+        constants.sector_size,
+        constants.message_size_max,
+    );
+    errdefer gpa.free(scrub.buffer_prepare);
+
+    return scrubbed_bytes;
+}
+
+fn open(scrub: *Scrub) !void {
+    scrub.client_sessions_checkpoint.open(
+        &scrub.grid,
+        scrub.superblock.working.client_sessions_reference(),
+        struct {
+            fn client_sessions_checkpoint_callback(_: *CheckpointTrailer) void {}
+        }.client_sessions_checkpoint_callback,
+    );
+    while (scrub.client_sessions_checkpoint.callback != .none) {
+        try scrub.io.run_for_ns(constants.tick_ms * std.time.ns_per_ms);
+    }
+    log.debug("client sessions checkpoint opened", .{});
+
+    scrub.grid.open(struct {
+        fn grid_open_callback(_: *Grid) void {}
+    }.grid_open_callback);
+    while (scrub.grid.callback != .none) {
+        try scrub.io.run_for_ns(constants.tick_ms * std.time.ns_per_ms);
+    }
+    log.debug("grid opened", .{});
+
+    scrub.forest.open(struct {
+        fn forest_open_callback(_: *Forest) void {}
+    }.forest_open_callback);
+    while (scrub.forest.progress != null) {
+        try scrub.io.run_for_ns(constants.tick_ms * std.time.ns_per_ms);
+    }
+    log.debug("forest opened", .{});
+}
+
+fn deinit(scrub: *Scrub, gpa: std.mem.Allocator) void {
+    gpa.free(scrub.buffer_headers);
+    gpa.free(scrub.buffer_prepare);
+
+    scrub.grid_blocks_scrubbed.deinit(gpa);
+    scrub.grid_scrubber.deinit(gpa);
+    scrub.forest.deinit(gpa);
+    scrub.grid.deinit(gpa);
+    scrub.client_sessions_checkpoint.deinit(gpa);
+    scrub.superblock.deinit(gpa);
+    scrub.storage.deinit();
+}
+
+/// Scrubs the WAL headers and prepares, using sync IO.
+fn scrub_wal(scrub: *Scrub) !u64 {
+    var scrubbed_bytes: u64 = 0;
+
+    const headers_bytes_read = try scrub.sync_read_all(
+        scrub.buffer_headers,
+        vsr.Zone.wal_headers.start(),
+    );
+    assert(headers_bytes_read == scrub.buffer_headers.len);
+
+    const wal_headers: []vsr.Header.Prepare align(16) = @alignCast(
+        std.mem.bytesAsSlice(vsr.Header.Prepare, scrub.buffer_headers),
+    );
+
+    for (wal_headers, 0..) |*wal_header, slot| {
+        const offset = slot * constants.message_size_max;
+
+        const bytes_read = try scrub.sync_read_all(
+            scrub.buffer_prepare,
+            vsr.Zone.wal_prepares.start() + offset,
+        );
+        assert(bytes_read == scrub.buffer_prepare.len);
+
+        const wal_prepare: *align(16) vsr.Header.Prepare = @alignCast(std.mem.bytesAsValue(
+            vsr.Header.Prepare,
+            scrub.buffer_prepare[0..@sizeOf(vsr.Header)],
+        ));
+
+        const wal_prepare_body_valid =
+            wal_prepare.valid_checksum() and
+            wal_prepare.valid_checksum_body(
+                scrub.buffer_prepare[@sizeOf(vsr.Header)..wal_prepare.size],
+            );
+
+        assert(wal_header.valid_checksum() and wal_prepare_body_valid);
```

Comment:
I'd split these so that if it hits this assertion, we have more information about the issue. (Also more readable, as a bonus.)

```suggestion
        assert(wal_header.valid_checksum());
        assert(wal_prepare_body_valid);
```